connector/src/main/java/org/geysermc/connector/network/translators/java/JavaDeclareRecipesTranslator.java

```diff
@@ -131,13 +143,65 @@ public void translate(ServerDeclareRecipesPacket packet, GeyserSession session)
                     craftingDataPacket.getCraftingData().addAll(RecipeRegistry.LEATHER_DYEING_RECIPES);
                     break;
                 }
+                case STONECUTTING: {
+                    StoneCuttingRecipeData stoneCuttingData = (StoneCuttingRecipeData) recipe.getData();
+                    ItemStack ingredient = stoneCuttingData.getIngredient().getOptions()[0];
+                    List<StoneCuttingRecipeData> data = unsortedStonecutterData.get(ingredient.getId());
+                    if (data == null) {
+                        data = new ArrayList<>();
+                        unsortedStonecutterData.put(ingredient.getId(), data);
+                    }
+                    data.add(stoneCuttingData);
+                    // Save for processing after all recipes have been received
+                }
             }
         }
+        // Add all cartography table recipe UUIDs, so we can use the cartography table
+        craftingDataPacket.getCraftingData().addAll(RecipeRegistry.CARTOGRAPHY_RECIPE_DATA);
+
         craftingDataPacket.getPotionMixData().addAll(PotionMixRegistry.POTION_MIXES);
+
+        Int2ObjectMap<IntList> stonecutterRecipeMap = new Int2ObjectOpenHashMap<>();
+        for (Int2ObjectMap.Entry<List<StoneCuttingRecipeData>> data : unsortedStonecutterData.int2ObjectEntrySet()) {
+            data.getValue().sort(Comparator.comparing((stoneCuttingRecipeData ->
+                    // Sort the list by each output item's Java identifier - this is how it's sorted on Java, and therefore
+                    // We can get the correct order for button pressing
```

Comment:
Move this to the first line of the for loop. That makes it more clear that the method ends.